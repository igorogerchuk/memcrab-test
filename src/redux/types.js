const types = {
  REMOVE_LINE: "REMOVE_LINE",
  ADD_LINE: "ADD_LINE",
  SAVE_ARRAY: "SAVE_ARRAY"
};

export default types;
