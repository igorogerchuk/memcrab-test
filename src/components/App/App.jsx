// @flow
import React from "react";
import Inputs from "../Inputs/Inputs";
import Array from "../Array";
import "./App.module.css";

const App = () => (
  <div styleName="container">
    <Inputs />
    <Array />
  </div>
);
export default App;
