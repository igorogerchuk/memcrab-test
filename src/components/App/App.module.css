html {
  box-sizing: border-box;
}

*,
*:before,
*:after {
  box-sizing: inherit;
  padding: 0;
  margin: 0;
}

body {
  background: linear-gradient(to right, #2980b9, #6dd5fa, #ffffff);
}

.container {
  width: 80%;
  margin: 0 auto;
  text-align: center;
  font-family: Arial, Helvetica, sans-serif;
}