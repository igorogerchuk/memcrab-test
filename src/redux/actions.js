// @flow

import * as types from "./types";
import type { ArrayState, RowsState, CellsState, ParamsState } from "./types";

type _ExtractReturn<R, F: (...args: any[]) => R> = R;
type ExtractReturn<F> = _ExtractReturn<*, F>;

export const removeRow = (id: string, cellsIds: $ReadOnlyArray<string>) => ({
  type: types.REMOVE_ROW,
  payload: { id, cellsIds }
});

export const addRow = (
  array: ArrayState,
  row: RowsState,
  cells: CellsState
) => ({
  type: types.ADD_ROW,
  payload: { array, row, cells }
});

export const saveArray = (array: ArrayState) => ({
  type: types.SAVE_ARRAY,
  payload: { array }
});

export const saveRows = (rows: RowsState) => ({
  type: types.SAVE_ROWS,
  payload: { rows }
});

export const saveCells = (cells: CellsState) => ({
  type: types.SAVE_CELLS,
  payload: { cells }
});

export const increase = (id: string) => ({
  type: types.INCREASE,
  payload: { id }
});

export const saveParams = (params: ParamsState) => ({
  type: types.SAVE_PARAMS,
  payload: { params }
});

export type Action =
  | ExtractReturn<typeof removeRow>
  | ExtractReturn<typeof addRow>
  | ExtractReturn<typeof saveArray>
  | ExtractReturn<typeof saveRows>
  | ExtractReturn<typeof saveCells>
  | ExtractReturn<typeof saveParams>
  | ExtractReturn<typeof increase>;
