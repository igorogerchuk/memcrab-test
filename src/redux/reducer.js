import types from "./types";

const reducer = (state = { array: [] }, { type, payload }) => {
  const { array } = state;
  switch (type) {
    case types.ADD_LINE:
      return {
        array: array.concat(payload.newLine)
      };
    case types.REMOVE_LINE:
      return {
        array: array.filter(line => line !== payload.line)
      };
    case types.SAVE_ARRAY:
      return { array: payload.array };
    default:
      return state;
  }
};

export default reducer;
