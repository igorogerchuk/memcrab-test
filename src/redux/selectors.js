// @flow
import { createSelector } from "reselect";
import type { State } from "./types";

export const getArray = (state: State) => state.array;

export const getRows = (state: State) => state.rows;

export const getCells = (state: State) => state.cells;

export const getColumnsQty = (state: State) => state.params.n;

export const getIlluminatedQty = (state: State) => state.params.x;

export const getSumColumn = createSelector<State, *, *, *, *>(
  [getRows, getCells],
  (rows, cells) => {
    const sumColumn = [];
    const cellsIds = Object.keys(rows).map((rowId: string) => rows[rowId]);
    for (let i = 0; i < cellsIds.length; i++) {
      let rowTotal = 0;
      for (let j = 0; j < cellsIds[i].length; j++) {
        rowTotal += cells[cellsIds[i][j]].amount;
      }
      sumColumn.push(rowTotal);
    }
    return sumColumn;
  }
);

export const getAvarageRow = createSelector<State, *, *, *, *>(
  [getRows, getCells],
  (rows, cells) => {
    const avarageRow = [];
    const cellsIds = Object.keys(rows).map((rowId: string) => rows[rowId]);
    for (let j = 0; j < cellsIds[0].length; j++) {
      let columnTotal = 0;
      for (let i = 0; i < cellsIds.length; i++) {
        columnTotal += cells[cellsIds[i][j]].amount;
      }
      avarageRow.push(columnTotal / cellsIds.length);
    }
    return avarageRow;
  }
);
