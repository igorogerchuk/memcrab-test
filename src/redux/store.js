// @flow
import { createStore } from "redux";
import reducer from "./reducer";
import type { State } from "./types";
import type { Dispatch } from "redux";
import type { Action } from "./actions";

const store = createStore<State, Action, Dispatch<Action>>(
  reducer,
  window.__REDUX_DEVTOOLS_EXTENSION__ && window.__REDUX_DEVTOOLS_EXTENSION__()
);
export default store;
