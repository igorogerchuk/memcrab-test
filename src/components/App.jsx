import React, { Component } from "react";
import Inputs from "./Inputs";
import Array from "./Array";
import AddLineButton from "./AddLineButton";
import { connect } from "react-redux";
import * as selectors from "../redux/selectors";

class App extends Component {
  render() {
    const { array } = this.props;
    return (
      <>
        <Inputs />
        {array.length > 0 && <Array />}
        {array.length > 0 && <AddLineButton />}
      </>
    );
  }
}

const mapStateToProps = state => {
  return { array: selectors.getArray(state) };
};

export default connect(mapStateToProps)(App);
