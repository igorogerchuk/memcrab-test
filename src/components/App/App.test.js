import React from "react";
import App from "./App";
import renderer from "react-test-renderer";
import { Provider } from "react-redux";
import { createStore } from "redux";
import reducer from "../../redux/reducer";
import createRandomArray from "../../services/randomArray";

it("should render correctly with initial state from Redux store", () => {
  const store = createStore(reducer);
  const snap = renderer
    .create(
      <Provider store={store}>
        <App />
      </Provider>
    )
    .toJSON();
  expect(snap).toMatchSnapshot();
});

const { array, rows, cells } = createRandomArray(5, 4);
const params = { n: 4, x: 3 };
const store = createStore(reducer, { array, rows, cells, params });

it("should render correctly with given random array 5x4", () => {
  const snap = renderer
    .create(
      <Provider store={store}>
        <App />
      </Provider>
    )
    .toJSON();
  expect(snap).toMatchSnapshot();
});
