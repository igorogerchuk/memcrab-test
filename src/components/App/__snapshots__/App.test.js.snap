// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly with given random array 5x4 1`] = `
<div
  className="container___2qaOc"
>
  <div
    className="formWrapper___Xr_Ln"
  >
    <h1
      className="formTitle___1Sczm"
    >
      Enter table parameters
    </h1>
    <form
      className="form___1t6C-"
      id="form"
      onSubmit={[Function]}
    >
      <div
        className="inputWrapper___1nGv5"
      >
        <label
          className="label___2sD9G"
          htmlFor="m"
        >
          Rows quantity:
        </label>
        <input
          className="input___3WuZL"
          id="m"
          name="rows"
          onChange={[Function]}
          placeholder="0"
          type="number"
          value=""
        />
      </div>
      <div
        className="inputWrapper___1nGv5"
      >
        <label
          className="label___2sD9G"
          htmlFor="n"
        >
          Columns quantity:
        </label>
        <input
          className="input___3WuZL"
          id="n"
          name="columns"
          onChange={[Function]}
          placeholder="0"
          type="number"
          value=""
        />
      </div>
      <div
        className="inputWrapper___1nGv5"
      >
        <label
          className="label___2sD9G"
          htmlFor="m"
        >
          Illuminated cells quantity:
        </label>
        <input
          className="input___3WuZL"
          id="x"
          name="illuminate"
          onChange={[Function]}
          placeholder="0"
          type="number"
          value=""
        />
      </div>
      <div>
        <input
          className="submitButton___1e-tC"
          type="submit"
          value="Generate"
        />
      </div>
    </form>
  </div>
  <div
    className="tableWrapper___3mKZT"
  >
    <table
      className="table___3XwP7"
    >
      <tbody>
        <tr
          id="1"
        >
          <td
            className="simpleTd___1ELya"
            id="2"
            onClick={[Function]}
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
            style={
              Object {
                "background": "linear-gradient(110deg,#eee 0%,#e3e3e3 10%,#fff 20%,#fff 25%,#f0f0f0 26%,#fff 28%,#ddd 55%,#eee 100%)",
              }
            }
          >
            500
          </td>
          <td
            className="simpleTd___1ELya"
            id="3"
            onClick={[Function]}
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
            style={
              Object {
                "background": "linear-gradient(110deg,#eee 0%,#e3e3e3 10%,#fff 20%,#fff 25%,#f0f0f0 26%,#fff 28%,#ddd 55%,#eee 100%)",
              }
            }
          >
            500
          </td>
          <td
            className="simpleTd___1ELya"
            id="4"
            onClick={[Function]}
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
            style={
              Object {
                "background": "linear-gradient(110deg,#eee 0%,#e3e3e3 10%,#fff 20%,#fff 25%,#f0f0f0 26%,#fff 28%,#ddd 55%,#eee 100%)",
              }
            }
          >
            500
          </td>
          <td
            className="simpleTd___1ELya"
            id="5"
            onClick={[Function]}
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
            style={
              Object {
                "background": "linear-gradient(110deg,#eee 0%,#e3e3e3 10%,#fff 20%,#fff 25%,#f0f0f0 26%,#fff 28%,#ddd 55%,#eee 100%)",
              }
            }
          >
            500
          </td>
          <td
            className="sumCell___2yW_R"
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
          >
            2000
          </td>
          <td
            className="removeButtonTd___1OWGN addCell___a54He"
          >
            <button
              className="removeButton___2Sb5_ addButton___2a67D"
              id="1"
              onClick={[Function]}
            >
              ×
            </button>
          </td>
        </tr>
        <tr
          id="6"
        >
          <td
            className="simpleTd___1ELya"
            id="7"
            onClick={[Function]}
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
            style={
              Object {
                "background": "linear-gradient(110deg,#eee 0%,#e3e3e3 10%,#fff 20%,#fff 25%,#f0f0f0 26%,#fff 28%,#ddd 55%,#eee 100%)",
              }
            }
          >
            500
          </td>
          <td
            className="simpleTd___1ELya"
            id="8"
            onClick={[Function]}
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
            style={
              Object {
                "background": "linear-gradient(110deg,#eee 0%,#e3e3e3 10%,#fff 20%,#fff 25%,#f0f0f0 26%,#fff 28%,#ddd 55%,#eee 100%)",
              }
            }
          >
            500
          </td>
          <td
            className="simpleTd___1ELya"
            id="9"
            onClick={[Function]}
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
            style={
              Object {
                "background": "linear-gradient(110deg,#eee 0%,#e3e3e3 10%,#fff 20%,#fff 25%,#f0f0f0 26%,#fff 28%,#ddd 55%,#eee 100%)",
              }
            }
          >
            500
          </td>
          <td
            className="simpleTd___1ELya"
            id="10"
            onClick={[Function]}
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
            style={
              Object {
                "background": "linear-gradient(110deg,#eee 0%,#e3e3e3 10%,#fff 20%,#fff 25%,#f0f0f0 26%,#fff 28%,#ddd 55%,#eee 100%)",
              }
            }
          >
            500
          </td>
          <td
            className="sumCell___2yW_R"
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
          >
            2000
          </td>
          <td
            className="removeButtonTd___1OWGN addCell___a54He"
          >
            <button
              className="removeButton___2Sb5_ addButton___2a67D"
              id="6"
              onClick={[Function]}
            >
              ×
            </button>
          </td>
        </tr>
        <tr
          id="11"
        >
          <td
            className="simpleTd___1ELya"
            id="12"
            onClick={[Function]}
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
            style={
              Object {
                "background": "linear-gradient(110deg,#eee 0%,#e3e3e3 10%,#fff 20%,#fff 25%,#f0f0f0 26%,#fff 28%,#ddd 55%,#eee 100%)",
              }
            }
          >
            500
          </td>
          <td
            className="simpleTd___1ELya"
            id="13"
            onClick={[Function]}
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
            style={
              Object {
                "background": "linear-gradient(110deg,#eee 0%,#e3e3e3 10%,#fff 20%,#fff 25%,#f0f0f0 26%,#fff 28%,#ddd 55%,#eee 100%)",
              }
            }
          >
            500
          </td>
          <td
            className="simpleTd___1ELya"
            id="14"
            onClick={[Function]}
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
            style={
              Object {
                "background": "linear-gradient(110deg,#eee 0%,#e3e3e3 10%,#fff 20%,#fff 25%,#f0f0f0 26%,#fff 28%,#ddd 55%,#eee 100%)",
              }
            }
          >
            500
          </td>
          <td
            className="simpleTd___1ELya"
            id="15"
            onClick={[Function]}
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
            style={
              Object {
                "background": "linear-gradient(110deg,#eee 0%,#e3e3e3 10%,#fff 20%,#fff 25%,#f0f0f0 26%,#fff 28%,#ddd 55%,#eee 100%)",
              }
            }
          >
            500
          </td>
          <td
            className="sumCell___2yW_R"
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
          >
            2000
          </td>
          <td
            className="removeButtonTd___1OWGN addCell___a54He"
          >
            <button
              className="removeButton___2Sb5_ addButton___2a67D"
              id="11"
              onClick={[Function]}
            >
              ×
            </button>
          </td>
        </tr>
        <tr
          id="16"
        >
          <td
            className="simpleTd___1ELya"
            id="17"
            onClick={[Function]}
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
            style={
              Object {
                "background": "linear-gradient(110deg,#eee 0%,#e3e3e3 10%,#fff 20%,#fff 25%,#f0f0f0 26%,#fff 28%,#ddd 55%,#eee 100%)",
              }
            }
          >
            500
          </td>
          <td
            className="simpleTd___1ELya"
            id="18"
            onClick={[Function]}
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
            style={
              Object {
                "background": "linear-gradient(110deg,#eee 0%,#e3e3e3 10%,#fff 20%,#fff 25%,#f0f0f0 26%,#fff 28%,#ddd 55%,#eee 100%)",
              }
            }
          >
            500
          </td>
          <td
            className="simpleTd___1ELya"
            id="19"
            onClick={[Function]}
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
            style={
              Object {
                "background": "linear-gradient(110deg,#eee 0%,#e3e3e3 10%,#fff 20%,#fff 25%,#f0f0f0 26%,#fff 28%,#ddd 55%,#eee 100%)",
              }
            }
          >
            500
          </td>
          <td
            className="simpleTd___1ELya"
            id="20"
            onClick={[Function]}
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
            style={
              Object {
                "background": "linear-gradient(110deg,#eee 0%,#e3e3e3 10%,#fff 20%,#fff 25%,#f0f0f0 26%,#fff 28%,#ddd 55%,#eee 100%)",
              }
            }
          >
            500
          </td>
          <td
            className="sumCell___2yW_R"
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
          >
            2000
          </td>
          <td
            className="removeButtonTd___1OWGN addCell___a54He"
          >
            <button
              className="removeButton___2Sb5_ addButton___2a67D"
              id="16"
              onClick={[Function]}
            >
              ×
            </button>
          </td>
        </tr>
        <tr
          id="21"
        >
          <td
            className="simpleTd___1ELya"
            id="22"
            onClick={[Function]}
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
            style={
              Object {
                "background": "linear-gradient(110deg,#eee 0%,#e3e3e3 10%,#fff 20%,#fff 25%,#f0f0f0 26%,#fff 28%,#ddd 55%,#eee 100%)",
              }
            }
          >
            500
          </td>
          <td
            className="simpleTd___1ELya"
            id="23"
            onClick={[Function]}
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
            style={
              Object {
                "background": "linear-gradient(110deg,#eee 0%,#e3e3e3 10%,#fff 20%,#fff 25%,#f0f0f0 26%,#fff 28%,#ddd 55%,#eee 100%)",
              }
            }
          >
            500
          </td>
          <td
            className="simpleTd___1ELya"
            id="24"
            onClick={[Function]}
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
            style={
              Object {
                "background": "linear-gradient(110deg,#eee 0%,#e3e3e3 10%,#fff 20%,#fff 25%,#f0f0f0 26%,#fff 28%,#ddd 55%,#eee 100%)",
              }
            }
          >
            500
          </td>
          <td
            className="simpleTd___1ELya"
            id="25"
            onClick={[Function]}
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
            style={
              Object {
                "background": "linear-gradient(110deg,#eee 0%,#e3e3e3 10%,#fff 20%,#fff 25%,#f0f0f0 26%,#fff 28%,#ddd 55%,#eee 100%)",
              }
            }
          >
            500
          </td>
          <td
            className="sumCell___2yW_R"
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
          >
            2000
          </td>
          <td
            className="removeButtonTd___1OWGN addCell___a54He"
          >
            <button
              className="removeButton___2Sb5_ addButton___2a67D"
              id="21"
              onClick={[Function]}
            >
              ×
            </button>
          </td>
        </tr>
        <tr>
          <td
            className="averageTd___FDFfU"
          >
            500.00
          </td>
          <td
            className="averageTd___FDFfU"
          >
            500.00
          </td>
          <td
            className="averageTd___FDFfU"
          >
            500.00
          </td>
          <td
            className="averageTd___FDFfU"
          >
            500.00
          </td>
        </tr>
        <tr>
          <td
            className="addCell___a54He"
            colSpan={4}
          >
            <button
              className="addButton___2a67D"
              id="addButton"
              onClick={[Function]}
            >
              +
            </button>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;

exports[`should render correctly with initial state from Redux store 1`] = `
<div
  className="container___2qaOc"
>
  <div
    className="formWrapper___Xr_Ln"
  >
    <h1
      className="formTitle___1Sczm"
    >
      Enter table parameters
    </h1>
    <form
      className="form___1t6C-"
      id="form"
      onSubmit={[Function]}
    >
      <div
        className="inputWrapper___1nGv5"
      >
        <label
          className="label___2sD9G"
          htmlFor="m"
        >
          Rows quantity:
        </label>
        <input
          className="input___3WuZL"
          id="m"
          name="rows"
          onChange={[Function]}
          placeholder="0"
          type="number"
          value=""
        />
      </div>
      <div
        className="inputWrapper___1nGv5"
      >
        <label
          className="label___2sD9G"
          htmlFor="n"
        >
          Columns quantity:
        </label>
        <input
          className="input___3WuZL"
          id="n"
          name="columns"
          onChange={[Function]}
          placeholder="0"
          type="number"
          value=""
        />
      </div>
      <div
        className="inputWrapper___1nGv5"
      >
        <label
          className="label___2sD9G"
          htmlFor="m"
        >
          Illuminated cells quantity:
        </label>
        <input
          className="input___3WuZL"
          id="x"
          name="illuminate"
          onChange={[Function]}
          placeholder="0"
          type="number"
          value=""
        />
      </div>
      <div>
        <input
          className="submitButton___1e-tC"
          type="submit"
          value="Generate"
        />
      </div>
    </form>
  </div>
</div>
`;
